sync/worker.go

```diff
@@ -0,0 +1,102 @@
+package sync
+
+import (
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p"
+	"github.com/spacemeshos/go-spacemesh/p2p/server"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type RequestFactory func(s *server.MessageServer, peer p2p.Peer) (chan interface{}, error)
+
+type worker struct {
+	*sync.Once
+	sync.WaitGroup
+	output chan interface{}
+	count  *int32
+	action func()
+	log.Log
+}
+
+func (w *worker) Work() {
+	w.Debug("worker work")
+	w.action()
+	atomic.AddInt32(w.count, -1)
+	w.Debug("worker done")
+	if atomic.LoadInt32(w.count) == 0 { //close once everyone is finished
+		w.Debug("worker teardown")
+		w.Do(func() { close(w.output) })
+	}
+}
+
+func NewWorker(s *server.MessageServer,
+	requestTimeout time.Duration,
+	peer p2p.Peer,
+	mu *sync.Once,
+	count *int32,
+	output chan interface{},
+	reqFactory RequestFactory) worker {
+	workFunc := func() {
+		log.Debug("send request Peer: %v", peer)
+		ch, _ := reqFactory(s, peer)
+		timeout := time.After(requestTimeout)
```

Comment:
from the docs of `time.After` - 
`If efficiency is a concern, use NewTimer instead and call Timer.Stop if the timer is no longer needed.`
not a must though.